File: Work_leetcode/Can I Win.cpp
class Solution {
private:
    unordered_map<int, bool> isUsed;
public:
    bool canIWin(int maxChoosableInteger, int desiredTotal) {
        bool ans = false;
        for (int i=1; i<=maxChoosableInteger; ++i) isUsed[i] = false;
        return helper(desiredTotal);
    }
    bool helper(int disiredTotal) {
        for (auto m : isUsed) {
            if (!m.second) {
                if (m.first > disiredTotal) return true;
                m.second = true;
                bool tempAns = helper(disiredTotal - m.first);
                if (!tempAns) return true;
                m.second = false;
            }
        }
        return false;
    }
};

File: Work_leetcode/Can I win.java
public class Solution {
    public boolean canIWin(int maxChoosableInteger, int desiredTotal) {
        List<Integer> candidates = new ArrayList<Integer>();
        for (int i = 1; i <= maxChoosableInteger; i++){
            candidates.add(i);
        }
        return helper(candidates, desiredTotal);
    }

    public boolean helper(List<Integer> candidates, int target){
        if (candidates.size() > 0 && candidates.get(candidates.size()-1) >= target){
            return true;
        }
        for (int i = 0; i < candidates.size(); i++){
            int removed = candidates.remove(i);
            if (!helper(candidates, target-removed)){
                candidates.add(i, removed);
                return true;
            }
            candidates.add(i, removed);
        }
        return false;
    }
}